### PR TITLE
Recipe App

### PR DESCRIPTION
**Pull request by PraisesPJMT**
## Project Link
- Project Kanban Link: [https://github.com/users/PraisesPJMT/projects/9](https://github.com/users/PraisesPJMT/projects/9)
- Repository Link: [https://github.com/PraisesPJMT/FoodPad-App](https://github.com/PraisesPJMT/FoodPad-App)

## Team Members
Here are the team members working on this project.
- Member 1: Paul Onchera (@Paul-js-hub)
- Member 2: Praises Tula (@PraisesPJMT)

## Project Description
Create a Recipe app keeps track of all your recipes, ingredients, and shopping list.

## Project Features
- We followed the layout of the wireframes provided. Some of the app features include:
- Login page and registration page:
    - Built with Devise.
- Food list:
    - Display a list of food added by the logged-in user as in the wireframe .
    - Leads to a form that allows users to add new food.
- Recipes list:
    - Displays a list of recipes created by the logged-in user as in the wireframe.
    - Leads to recipe details.
    - If the user is the owner of the recipe, allow the user to delete it.
- Public recipe list:
    - Displays a list of all public recipes ordered by newest as in the wireframe.
    - Leads to recipe details.
- Recipe details:
    - If the recipe is public or the user is the owner of the recipe, displays the recipe details as in the wireframe.
    - If the user is the owner of the recipe, leads to the form that allows the user to add new food.
- General shopping list view :
    - Show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
    - Count the total food items and total price of the missing food.
- Made sure there are no N+1 queries happening.
- Created a navigation menu that allows users to open all of the pages created.

Happy Reviewing!!!!

### CODE REVIEW

File: app/models/ability.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    #
+    # See the wiki for details:
+    # https://github.com/CanCanCommunity/cancancan/blob/develop/docs/define_check_abilities.md
+    user ||= User.new
+
+    can :read, :all
+    can :create, :all
+    can(:destroy, Food, user:)
+    can(:destroy, Recipe, user:)
+    can(:destroy, RecipeFood, user:)
```
Comment (Munsa1) [inline app/models/ability.rb]: - [x] I am get a syntax error when I try to add a food, kindly fix this
image

File: app/models/recipe_food.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class RecipeFood < ApplicationRecord
+  belongs_to :recipe
+  belongs_to :food
```
Comment (Munsa1) [inline app/models/recipe_food.rb:3]: - [ ] Again when trying to create a new recepie I get the following syntax error, kindly fix it.
image

File: app/views/devise/confirmations/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend confirmation instructions</h2>
+
+<%= form_for(resource, as: resource_name, url: confirmation_path(resource_name), html: { method: :post }) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true, autocomplete: "email", value: (resource.pending_reconfirmation? ? resource.unconfirmed_email : resource.email) %>
+  </div>
+
```
Comment (Munsa1) [inline app/views/devise/confirmations/new.html.erb:10]: - [x] Kindly ensure that the devise login page disappears upon successful login
image

File: app/views/foods/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<h1>New food</h1>
+
```
Comment (Munsa1) [inline app/views/foods/new.html.erb]: - [ ] Kindly ensure that you follow the design templates for all the views as this is a requirement. This means following how components are displayed in all the given designs to match with your designs. [Link](https://github.com/microverseinc/curriculum-rails/blob/main/recipe-app/buisness_requirements.md)

Recepies
image

image

image

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @PraisesPJMT  @Paul-js-hub , 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>


## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: app/views/devise/confirmations/new.html.erb (hunk 3 above)
Comment (PraisesPJMT) [reply app/views/devise/confirmations/new.html.erb:10]: This feature has now been implemented! Thanks so much for the review!

File: app/views/foods/new.html.erb (hunk 4 above)
Comment (PraisesPJMT) [reply app/views/foods/new.html.erb]: Updated all view to follow wireframe! Thanks so much for the review!

File: app/models/ability.rb (hunk 1 above)
Comment (PraisesPJMT) [reply app/models/ability.rb]: Modified the authorization for user! Thanks so much for the feedback!

File: app/models/recipe_food.rb (hunk 2 above)
Comment (PraisesPJMT) [reply app/models/recipe_food.rb:3]: Modified the authorization for user! Thanks so much for the feedback!

File: app/models/ability.rb (hunk 1 above)
Comment (PraisesPJMT) [reply app/models/ability.rb]: Modified the authorization for user! Thanks so much for the feedback!

**Comment by PraisesPJMT** [comment]
## Review Feedback
Thanks so much @Munsa1  for the amazing feedback.
## Review Changes
The following changes were made based on the review:
- [Fix route for page after log-in](https://github.com/PraisesPJMT/FoodPad-App/pull/20/commits/44b981a388dd1195b3eba6a13d4ace0cadc4089a)
- [Adjust view to follow wireframe](https://github.com/PraisesPJMT/FoodPad-App/pull/20/commits/524f685e4c9659bd755f1324be2e9558418a88f7)
- [Update user authorization](https://github.com/PraisesPJMT/FoodPad-App/pull/20/commits/16637654729d9fc85644a3afea211c3f2b0d8558)


Happy Reviewing!!!!!!

File: db/migrate/20221220085830_add_devise_to_users.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      # t.datetime :current_sign_in_at
+      # t.datetime :last_sign_in_at
+      # t.string   :current_sign_in_ip
+      # t.string   :last_sign_in_ip
+
+      ## Confirmable
+      # t.string   :confirmation_token
+      # t.datetime :confirmed_at
+      # t.datetime :confirmation_sent_at
+      # t.string   :unconfirmed_email # Only if using reconfirmable
```
Comment (Ismailco) [inline db/migrate/20221220085830_add_devise_to_users.rb]: - [x] [OPTIONAL] When I create a new account, I got an error kindly make sure to uncomment the lines of code from 25 to28 to fix this 

Screenshot_20221224_094919

**Review by Ismailco — APPROVED** [review]
### Hi @PraisesPJMT, @Paul-js-hub,

In my opinion, your project is ready for presentation!  There is nothing else to say other than... it's time to merge it :shipit: 

cr-approve

## Highlights

✔️  No linter errors
✔️ Used Gitflow
✔️ Work documented professionally

Good luck! 🎉 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

File: db/migrate/20221220085830_add_devise_to_users.rb (hunk 5 above)
Comment (PraisesPJMT) [reply db/migrate/20221220085830_add_devise_to_users.rb]: Noted! We just fixed the error! nThanks so much for the feedback!

**Comment by PraisesPJMT** [comment]
Thanks so much @Ismailco  and @Munsa1 for the amazing review and feedback! 


Happy Codding!!!!!

**Pull request merged**